be1-go/channel/authentication/mod.go

```diff
@@ -0,0 +1,435 @@
+		latestRollCallMembers: newRollCallSet(),
+		hub:                   hub,
+		log:                   log,
+	}
+
+	newChannel.registry = newChannel.NewAuthenticationRegistry()
+
+	return newChannel
+}
+
+// ---
+// Publish-subscribe / channel.Channel implementation
+// ---
+
+// Subscribe : for authentication messages, we explicitly forbid subscription from clients.
+func (c *Channel) Subscribe(_ socket.Socket, _ method.Subscribe) error {
+	return xerrors.New("It is not possible to subscribe to the authentication channel.")
+}
+
+// Unsubscribe is also not usable in that context, as clients can't use Subscribe
+func (c *Channel) Unsubscribe(_ string, _ method.Unsubscribe) error {
+	return xerrors.New("It is not possible to unsubscribe from the authentication channel.")
+}
+
+// Publish is used to handle publish messages in the authentication channel.
+func (c *Channel) Publish(publish method.Publish, socket socket.Socket) error {
+	c.log.Info().
+		Str(msgID, strconv.Itoa(publish.ID)).
+		Msg("received a publish")
+
+	err := c.verifyMessage(publish.Params.Message)
+	if err != nil {
+		return xerrors.Errorf("failed to verify publish message on a "+
+			"authentication channel: %w", err)
+	}
+
+	err = c.handleMessage(publish.Params.Message, socket)
+	if err != nil {
+		return xerrors.Errorf("failed to handle publish message: %v", err)
+	}
+
+	return nil
+}
+
+// Catchup is used to handle a catchup message.
+func (c *Channel) Catchup(catchup method.Catchup) []message.Message {
+	c.log.Info().
+		Str(msgID, strconv.Itoa(catchup.ID)).
+		Msg("received a catchup")
+
+	return c.inbox.GetSortedMessages()
+}
+
+// Broadcast is forbidden, as authentication messages must be kept secret.
+func (c *Channel) Broadcast(_ method.Broadcast, _ socket.Socket) error {
+	panic("Broadcasting is not allowed in the authentication channel")
+}
+
+// ---
+// Message handling
+// ---
+
+// handleMessage handles a message received in a broadcast or publish method
+func (c *Channel) handleMessage(msg message.Message, socket socket.Socket) error {
+	err := c.registry.Process(msg, socket)
+	if err != nil {
+		return xerrors.Errorf("failed to process message: %w", err)
+	}
+
+	c.inbox.StoreMessage(msg)
+
+	return nil
+}
+
+// NewAuthenticationRegistry creates a new registry for an authentication channel
+func (c *Channel) NewAuthenticationRegistry() registry.MessageRegistry {
+	newRegistry := registry.NewMessageRegistry()
+	newRegistry.Register(messagedata.AuthenticateUser{}, c.auhenticateUser)
+	return newRegistry
+}
+
+func (c *Channel) auhenticateUser(msg message.Message, msgData interface{},
+	_ socket.Socket) error {
+
+	data, ok := msgData.(*messagedata.AuthenticateUser)
+	if !ok {
+		return xerrors.Errorf("message %v isn't a authentication message", msgData)
+	}
+
+	// verify signatures and validity of the pop Token.
+	err := c.verifyAuthMessage(msg, *data)
+	if err != nil {
+		c.log.Err(err).Msg("Error: invalid authentication message received")
```

Comment:
Is it necessary to both log and return an error here ?

Reply:
not really, it was more for debug purposes, i'll change that!